server/src/main/java/org/elasticsearch/search/rank/feature/RankFeatureShardPhase.java

```diff
@@ -0,0 +1,105 @@
+import org.elasticsearch.search.SearchHits;
+import org.elasticsearch.search.fetch.FetchSearchResult;
+import org.elasticsearch.search.fetch.StoredFieldsContext;
+import org.elasticsearch.search.fetch.subphase.FetchFieldsContext;
+import org.elasticsearch.search.fetch.subphase.FieldAndFormat;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.search.rank.context.RankFeaturePhaseRankShardContext;
+import org.elasticsearch.tasks.TaskCancelledException;
+
+import java.util.Arrays;
+import java.util.Collections;
+
+/**
+ * The {@code RankFeatureShardPhase} executes the rank feature phase on the shard, iff there is a {@code RankBuilder} that requires it.
+ * This phase is responsible for reading field data for a set of docids. To do this, it reuses the {@code FetchPhase} to read the required
+ * fields for all requested documents using the `FetchFieldPhase` sub-phase.
+ */
+public final class RankFeatureShardPhase {
+
+    private static final Logger LOGGER = LogManager.getLogger(RankFeatureShardPhase.class);
+
+    public static final RankFeatureShardResult EMPTY_RESULT = new RankFeatureShardResult(new RankFeatureDoc[0]);
+
+    public RankFeatureShardPhase() {}
+
+    public void prepareForFetch(SearchContext searchContext, RankFeatureShardRequest request) {
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("{}", new SearchContextSourcePrinter(searchContext));
+        }
+
+        if (searchContext.isCancelled()) {
+            throw new TaskCancelledException("cancelled");
+        }
+
+        RankFeaturePhaseRankShardContext rankFeaturePhaseRankShardContext = shardContext(searchContext);
+        if (rankFeaturePhaseRankShardContext != null) {
+            assert rankFeaturePhaseRankShardContext.getField() != null : "field must not be null";
+            searchContext.fetchFieldsContext(
+                new FetchFieldsContext(Collections.singletonList(new FieldAndFormat(rankFeaturePhaseRankShardContext.getField(), null)))
+            );
+            searchContext.storedFieldsContext(StoredFieldsContext.fromList(Collections.singletonList(StoredFieldsContext._NONE_)));
+            searchContext.addFetchResult();
+            Arrays.sort(request.getDocIds());
+        }
+    }
+
+    public void processFetch(SearchContext searchContext) {
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("{}", new SearchContextSourcePrinter(searchContext));
+        }
+
+        if (searchContext.isCancelled()) {
+            throw new TaskCancelledException("cancelled");
+        }
+
+        RankFeaturePhaseRankShardContext rankFeaturePhaseRankShardContext = searchContext.request().source().rankBuilder() != null
+            ? searchContext.request().source().rankBuilder().buildRankFeaturePhaseShardContext()
+            : null;
+        if (rankFeaturePhaseRankShardContext != null) {
+            RankFeatureShardResult featureRankShardResult = null;
+            try {
```

Comment:
I don't understand the `try {...} finally {...}`. There is no exception handling. Why not simply set the value after its constructed?

Reply:
There was some minor error handling ( + some dec-ref for `SearchHits` which wasn't needed after all), but moved around a few things and this was left out :/
Will remove the unnecessary try-finally block.